language/move-vm/runtime/src/loader.rs

```diff
@@ -1367,6 +1378,58 @@ impl Loader {
             }
         }
     }
+
+    // Check instantiation operation: count the # of nodes in `type_` and instantiate
+    // over ty_args. In practice, mimics instantiation without perfomring it.
+    // If that number is larger than `max_count`, refuse to construct this type.
+    // This prevents constructing larger and lager types via instantiation operations.
+    pub(crate) fn check_type_instantiation(
+        &self,
+        type_: &Type,
+        ty_args: &[Type],
+        max_count: usize,
+    ) -> PartialVMResult<usize> {
+        let mut type_params_count = vec![0; ty_args.len()];
+        let mut count_inst = self.count_type_nodes_and_params(type_, &mut type_params_count)?;
+        if count_inst > max_count {
+            return Err(PartialVMError::new(StatusCode::TOO_MANY_TYPE_NODES));
+        }
+        for (idx, count) in type_params_count.into_iter().enumerate() {
+            if count > 0 {
+                count_inst += count
+                    * match ty_args.get(idx) {
```

Comment:
`* match` with this big, multi-line match made me double take. Totally fine after reading it, but might be more clear if it was bound to its own local. 
Not a big deal either way, just wanted to share that (it is also a bit late for me fwiw... so maybe thats part of it)

Reply:
very true and I may do that, I was just so sick of it after my 14th attempt....

Reply:
done